crates/fuel-core/src/query/balance.rs

```diff
@@ -36,22 +40,28 @@ impl ReadView {
         asset_id: AssetId,
         base_asset_id: AssetId,
     ) -> StorageResult<AddressBalance> {
-        let amount = AssetQuery::new(
-            &owner,
-            &AssetSpendTarget::new(asset_id, u64::MAX, u16::MAX),
-            &base_asset_id,
-            None,
-            self,
-        )
-        .coins()
-        .map(|res| res.map(|coins| coins.amount()))
-        .try_fold(0u64, |balance, amount| {
-            async move {
-                // Increase the balance
-                Ok(balance.saturating_add(amount))
-            }
-        })
-        .await?;
+        let amount = if self.balances_enabled {
+            debug!(%owner, %asset_id, "Querying balance with balances cache");
+            self.off_chain.balance(&owner, &asset_id, &base_asset_id)?
+        } else {
+            debug!(%owner, %asset_id, "Querying balance without balances cache");
+            AssetQuery::new(
+                &owner,
+                &AssetSpendTarget::new(asset_id, u64::MAX, u16::MAX),
+                &base_asset_id,
+                None,
+                self,
+            )
+            .coins()
+            .map(|res| res.map(|coins| coins.amount()))
+            .try_fold(0u64, |balance, amount| {
+                async move {
+                    // Increase the balance
+                    Ok(balance.saturating_add(amount))
```

Comment:
Also not related to this PR, but is it possible for somebody to have more than `u64::MAX` amount of a single base_asset?

Reply:
I think it's possible, since balances on Ethereum are represented with 256 byte integers. While we cut 9 decimals when bridging to Fuel, that's still not enough to cover the full range of possible values with a `u64`.

Reply:
I didn't want to change [the type](https://github.com/FuelLabs/fuel-core/blob/master/crates/types/src/services/graphql_api.rs#L14) used in the original query. We might consider a follow-up PR to update the type used for balance amounts.